packages/workbox-window/Workbox.mjs

```diff
@@ -0,0 +1,417 @@
+    this._onStateChange = this._onStateChange.bind(this);
+    this._onUpdateFound = this._onUpdateFound.bind(this);
+  }
+
+  /**
+   * Registers a service worker for this instances script URL and service
+   * worker options. By default this method delays registration until after
+   * the window has loaded.
+   *
+   * @param {Object} [options]
+   * @param {Function} [options.immediate=false] Setting this to true will
+   *     register the service worker immediately, even if the window has
+   *     not loaded (not recommended).
+   */
+  async register({immediate = false} = {}) {
+    if (process.env.NODE_ENV !== 'production') {
+      if (this._registrationTime) {
+        logger.error('Cannot re-register a Workbox instance after it has ' +
+            'been registered. Create a new instance instead.');
+        return;
+      }
+    }
+
+    if (!immediate && document.readyState !== 'complete') {
+      await new Promise((res) => addEventListener('load', res));
+    }
+
+    // Create a local reference for better minification/
+    let reg;
+
+    try {
+      reg = await navigator.serviceWorker.register(
+          this._scriptUrl, this._swOptions);
+    } catch (error) {
+      if (process.env.NODE_ENV !== 'production') {
+        logger.error(error);
+      }
+      // Re-throw the error.
+      throw error;
+    }
+
+    // Determine if there's already an active SW with the same script URL.
+    // If there is, set it on the instance in the event that the registration
+    // doesn't trigger an update.
+    const activeSw = reg.active;
+    if (activeSw && urlsMatch(activeSw.scriptURL, this._scriptUrl)) {
+      this._sw = activeSw;
+      this._swPromiseResolver(activeSw);
+    }
+
+    // We set this at registration time rather than instantiation time
+    // in the unlikely event that a page creates multiple Workbox instances
+    // but staggers their registration.
+    this._controllingSw = navigator.serviceWorker.controller;
+
+    // If a SW with the same URL was already controlling the page, let it know
+    // the window is ready to receive messages.
+    if (this._controllingSw &&
+        urlsMatch(this._controllingSw.scriptURL, this._scriptUrl)) {
+      this._controllingSw.postMessage({
+        type: 'WINDOW_READY',
+        meta: 'workbox-window',
+      });
+    }
+
+    if (process.env.NODE_ENV !== 'production') {
+      logger.log('Successfully registered service worker.');
+
+      if (this._controllingSw) {
+        if (urlsMatch(this._controllingSw.scriptURL, this._scriptUrl)) {
+          logger.debug('A service worker with the same script URL is ' +
+              'already controlling this page. This service worker will ' +
+              'remain active unless an update is found.');
+        } else {
+          logger.debug('A service worker with a different script URL is ' +
+              'currently controlling the page.');
+          logger.debug('Fetching the new script...');
```

Comment:
What does this refer to? A request that our library is making, or the fact that the browser is implicitly fetching the new SW script for us?

Reply:
The latter. I could make it more obvious by saying "Fetching the new service worker script...". WDYT?

Reply:
How about `'The browser is fetching the new service worker script...'`. The main thing to get across is that `workbox-window` is just logging to provide info about something that the browser is doing anyway.